### PR TITLE
Add test for index links on dynamic routes.

### PR DESCRIPTION
**Pull request by Agrosis**
@mjackson, so this is what I've got for a test #2072. 

I didn't see any tests for dynamic routes and I didn't know where it should go best, so I went ahead and created a new file for that, as well as a directory for the routes that should be dynamically loaded. Lemme know if that's wrong or you've got something better in mind!

And second, I couldn't actually test this, I'm assuming that its because dynamic routing is done through webpack. How should I actually make sure the test works?

### CODE REVIEW

**Comment by mjackson** [comment]
Thanks, @Agrosis! We have some tests for async route configs in [here](https://github.com/rackt/react-router/blob/e7c6f3d848e55dda11595447928e843d39bed0eb/modules/__tests__/matchRoutes-test.js#L175). Basically, what you need to do is setup a route that uses an async `getChildRoutes`, that's all. You don't need to use `require.ensure` or webpack, just [use `setTimeout` to simulate async behavior](https://github.com/rackt/react-router/blob/e7c6f3d848e55dda11595447928e843d39bed0eb/modules/__tests__/matchRoutes-test.js#L184).

**Comment by Agrosis** [comment]
@mjackson , so I did some of what you said, but I think I'm running into a roadblock when I pass in the routes configuration to a Router component. Once the Router is rendered, I don't think there's enough time for the async route to be completely loaded before the test ends.

Should I even be using the Router component at all? I was assuming I had to render an IndexLink to make sure its active.

Sorry for taking so much time ):

File: modules/__tests__/dynamicRoutes/CheckoutPage.js [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import { Component } from 'react'
```
Comment (mjackson) [inline modules/__tests__/dynamicRoutes/CheckoutPage.js]: No need for a separate file for each component. Please just put it in the same file like [this](https://github.com/rackt/react-router/blob/e7c6f3d848e55dda11595447928e843d39bed0eb/modules/__tests__/Router-test.js#L12-L22)

Comment (Agrosis) [inline modules/__tests__/dynamicRoutes/CheckoutPage.js]: Oh, I fixed that in the 2nd commit, that was an outdated diff.

**Comment by mjackson** [comment]
You're almost there, @Agrosis! :)

Since the routes aren't loaded until later, you can't make your assertions on the initial render (as you've already discovered). The pattern I've been using to test async steps can be seen in a few places in the tests like [here](https://github.com/rackt/react-router/blob/e7c6f3d848e55dda11595447928e843d39bed0eb/modules/__tests__/Link-test.js#L124-L134) and [here](https://github.com/rackt/react-router/blob/e7c6f3d848e55dda11595447928e843d39bed0eb/modules/__tests__/Link-test.js#L299-L307).

For this test, we need to use the `<Router>`'s `onUpdate` hook to know when it has loaded its async routes and we can make the assertion that the link is active.

**Comment by Agrosis** [comment]
@mjackson I think I got it! So, at first I just link to a normal child route, and then in the first step I make sure that the child route is active and the index route isn't, then I push the state to the index route, and make sure the vice versa.

Any other changes you think I need here?

**Comment by knowbody** [comment]
@Agrogis any chance we could merge that? mind rebasing?

**Comment by Agrosis** [comment]
@knowbody done!

**Comment by knowbody** [comment]
so this is the failing test case for #2072 right?

**Comment by Agrosis** [comment]
@knowbody that's the one

**Comment by knowbody** [comment]
cool I'll look into it, thanks again :)

**Comment by taion** [comment]
Incorporated into #2425, which fixes the bug.

**Pull request closed without merging**